### PR TITLE
Add Target::LegacyBufferWrappers feature

### PR DESCRIPTION
**Pull request by steven-johnson**
This disables the default generation of the classic 'buffer_t' wrapper code for AOT compilation; the code is still present, but must be deliberately opted in by using the new Target::LegacyBufferWrappers flag when compiling. If you need to enable this flag for your code, you should strongly consider upgrading it to use halide_buffer_t instead (perhaps via the Halide::Runtime::Buffer<> class), as support for the classic 'buffer_t' struct will likely be removed at some point in the future.

### CODE REVIEW

File: src/LLVM_Runtime_Linker.cpp [hunk 1]
```diff
@@ -744,9 +744,15 @@ std::unique_ptr<llvm::Module> get_initial_module_for_target(Target t, llvm::LLVM
...
             modules.push_back(get_initmod_float16_t(c, bits_64, debug));
-            modules.push_back(get_initmod_old_buffer_t(c, bits_64, debug));
             modules.push_back(get_initmod_errors(c, bits_64, debug));
 
+
+            // Note that we deliberately include this module, even if Target::LegacyWrappers
+            // isn't enabled: it isn't much code, and it makes it much easier to
+            // intermingle code that is built with this flag with code that is
+            // built without.
+            modules.push_back(get_initmod_old_buffer_t(c, bits_64, debug));
```
Comment (abadams) [inline src/LLVM_Runtime_Linker.cpp]: define_extern calls to functions that accept old buffer_t's still need this module too

**Comment by abadams** [comment]
is "legacy_wrappers" too vague? Maybe "legacy_buffer_wrappers"?

**Comment by steven-johnson** [comment]
> is "legacy_wrappers" too vague? Maybe "legacy_buffer_wrappers"?

"legacy_buffer_wrappers_yes_this_is_deliberately_long_and_inconvenient_please_update_your_code"

File: Makefile [hunk 2]
```diff
@@ -1132,7 +1132,12 @@ $(FILTERS_DIR)/cxx_mangling_define_extern.a: $(BIN_DIR)/cxx_mangling_define_exte
 	$(CURDIR)/$< -g cxx_mangling_define_extern $(GEN_AOT_OUTPUTS) -o $(CURDIR)/$(FILTERS_DIR) target=$(TARGET)-no_runtime-c_plus_plus_name_mangling-user_context -f "HalideTest::cxx_mangling_define_extern"
 	$(ROOT_DIR)/tools/makelib.sh $@ $@  $(FILTERS_DIR)/cxx_mangling_define_extern_externs.o
 
-# pyramid needs a custom arg
+# Needs legacy_buffer_wrappers in the target.
```
Comment (zvookin) [inline Makefile]: I would change this comment to state that it actually tests the feature gated on the flag.

File: util/HalideTraceConfig.h [hunk 3]
```diff
@@ -0,0 +1,117 @@
+#ifndef HALIDE_TRACE_CONFIG_H
```
Comment (zvookin) [inline util/HalideTraceConfig.h]: Is this supposed to be here?

**Comment by shoaibkamil** [comment]
+1.  We transitioned to `halide_buffer_t` in our codebase a while back, and don't rely on legacy `buffer_t` support.

File: src/WrapExternStages.cpp [hunk 4]
```diff
@@ -145,6 +145,10 @@ class WrapExternStages : public IRMutator2 {
 }
 
 void wrap_legacy_extern_stages(Module m) {
+    if (!m.target().has_feature(Target::LegacyBufferWrappers)) {
```
Comment (abadams) [inline src/WrapExternStages.cpp]: This removes the ability to call extern stages that request the old buffer_t (the uses_old_buffer_t argument to define_extern). Bailing out here would make it a mysterious error. We could assert that the uses_old_buffer_t argument is false unless this target flag is on to push people to upgrade their extern stages, or we could just continue to wrap those stages - those wrappers currently only get emitted if you explicitly request one, so (unlike the global wrapper for the pipeline code) they're harmless in terms of code size.

Comment (abadams) [reply src/WrapExternStages.cpp]: Maybe in WrapExternStages::visit(const Call *) - that's when you discover if a call node has the uses_old_buffer_t flag.

Comment (steven-johnson) [reply src/WrapExternStages.cpp]: yep, testing now. (fwiw, it looks like we may not have *any* tests of `define_extern(use_old_buffer_t=true)` in Halide at the moment.)

Comment (steven-johnson) [reply src/WrapExternStages.cpp]: Correction: generator_aot_old_buffer_t tests this case.

**Comment by steven-johnson** [comment]
Think we're good here. PTAL.

**Pull request merged**